guava/src/com/google/common/util/concurrent/Uninterruptibles.java

```diff
@@ -326,7 +326,67 @@ public static boolean tryAcquireUninterruptibly(
     }
   }
 
-  // TODO(user): Add support for waitUninterruptibly.
+  /**
+   * Invokes {@code object.}{@link Object#wait() wait()} uninterruptibly.
+   *
+   * @param object to wait upon.
+   *
+   * @throws IllegalMonitorStateException if the current thread is not
+   *               the owner of the object's monitor.
+   */
+  @GwtIncompatible // concurrency
+  public static void waitUninterruptibly(Object object) {
+    Preconditions.checkNotNull(object);
+    boolean interrupted = false;
+    try {
+      while (true) {
+        try {
+          object.wait();
+          return;
+        } catch (InterruptedException e) {
+          interrupted = true;
```

Comment:
Interesting. I completely overlooked that difference. I am fine with the code which you shared. But wanted to discuss a possibility:

In the code which you shared, the usage pattern of the users of this API would be something like this:

```
while(!checkNotificationCondition()){
  Uninterruptibles.waitUninterruptibly(waitOn);
}
```

if we could change the API to something like this: 
```
public static void waitUninterruptibly(Object waitOn, Supplier<Boolean> notificationCondition) {
  boolean reinterrupt = Thread.interrupted();
  while(!notificationCondition.get()){
    try {
      waitOn.wait();
    } catch (InterruptedException e) {
      reinterrupt = true;
    }
  }
  if (reinterrupt) {
    selfInterrupt();
  }
}
```

And the usage will look like this:

```
Uninterruptibles.waitUninterruptibly(waitOn, this::notificationCondition())
```

Your thoughts?